Remove underline from button blocks

Aljullu opened the pull request:
See related issue #1285

The underline was intended to be added to links that look like links.

### Screenshots
_Before:_
![imatge](https://user-images.githubusercontent.com/3616980/79196321-2905d700-7e30-11ea-8bc8-d6762c8ca41f.png)

_After:_
![imatge](https://user-images.githubusercontent.com/3616980/79196278-125f8000-7e30-11ea-9544-f6cf038d5f95.png)

### How to test the changes in this Pull Request:

1. Add a _Buttons_ block.
2. In the frontend, verify they don't have an underline.

### Changelog

> Buttons blocks no longer show an underline.

haszari commented on assets/css/base/_base.scss:
```diff
@@ -1556,7 +1556,7 @@ button.menu-toggle {
 
-		a:not( .button ):not( .components-button ) {
+		a:not( .button ):not( .components-button ):not( .wp-block-button__link ) {
 			@include underlinedLink();
```
These rules all look backwards to me – can you add a comment about the intention for each?

Aljullu commented:
I refactored this a bit to make code clearer and easier to reuse, it also decreases selector specificity. I did a round of testing and everything was looking good to me. But let me know if you catch something I missed.

> In my testing this fixes the button issue, but introduces a regression for regular text links (not a button) - these homepage links now have underline (it disappears on mouseover though grin) (featured categories, new products, also affects all products block):

Good catch! That was a specificity issue, should be fixed with the refactor.

> By the way, I there is an option for button component to render as a link – would be good to test that use case also if possible. That might need a custom block/plugin to test (not sure of the details).

Added an specific not selector for it: b2f9020db919843875944fec9f07f1637f52307c.

haszari commented on assets/css/sass/utils/_mixins.scss:
```diff
@@ -32,6 +32,15 @@
 }
 
+// Remove underline if element has a button class.
+@mixin notUnderlinedWhenButton() {
```
I wonder if we can make this name clearer - starting with `not` always confuses me. Not a blocker but if you come up with something clearer before merge go for it :)

Does this remove underline from buttons? If so, we could call this `removeButtonUnderline` or similar.

Aljullu replied:
Yeah, much better name. Renamed in 97894ccfa67b3367a7535117b89afb7a1e978f6d.

haszari commented on assets/css/sass/utils/_mixins.scss:
```diff
@@ -32,6 +32,15 @@
+// Remove underline if element has a button class.
+@mixin notUnderlinedWhenButton() {
+	&.button,
+	&.components-button:not( .is-link ),
```
`stylelint` complained about this so I fixed it ☑️ b1c3ef31ef6a5f6abfb00912870016a9ad83cdff

```
assets/css/sass/utils/_mixins.scss
 38:26  ✖  Expected single space after "("    selector-pseudo-class-parentheses-space-inside
 38:33  ✖  Expected single space before ")"   selector-pseudo-class-parentheses-space-inside
```

Aljullu replied:
Thank you! :tada:

haszari commented on assets/css/base/_base.scss:
```diff
@@ -783,12 +785,14 @@ table {
 				text-decoration: none;
 			}
+
+			@include notUnderlinedWhenButton;
```
If we typically add the `()` to all mixins might be nice to do so here.

Aljullu replied:
Done in 97894ccfa67b3367a7535117b89afb7a1e978f6d.
